packages/service-library/src/servicelib/redis.py

```diff
@@ -135,7 +135,10 @@ async def _extend_lock(lock: Lock) -> None:
                 # lock is in use now
                 yield ttl_lock
         finally:
-            await ttl_lock.release()
+            # NOTE: lock can expire before being released.
+            # Will raise LockNotOwnedError
+            with contextlib.suppress(redis.exceptions.LockNotOwnedError):
```

Comment:
Question here: this issue happens only in the case of the multiple locks?
is that due to the fact that you release millions of times? thousands? hundreds?
in this case it could very well happen as well in a real use-case at some point?
like this it will go silent and we will not realize there is an issue right?

Reply:
The situation is the following:
- 250 locks are acquire in parallel with the option `blocking=True`, meaning: it will wait for the lock to be free before acquiring it.
- when the lock is acquired the `_extend_lock` task is started in the background extending the lock at a fixed interval of time, which is half of the duration of the lock's TTL
- before the task is released the lock extension is cancelled
Here is where the issue occurs:
- some time passes between the task's cancellation and the call to release
- if the TTL is too small, 1/2 of the TTL might be just shorter than the time it passes to between the task is canceled and the task lock is released
- this means that the lock will expire and be considered as not owned any longer

I think this is precisely the situation we are in.

For this test the TTL is set to `0.25` seconds, and half of that is `0.125` seconds

Reply:
Great! Can you please add this detail explanation in the code?

Reply:
What is implied above is that only one "task" owns and extends the task at a time. So the issue apparently is related to some timing.

Reply:
ok, so why is the time between task cancellation and the call to release to long? what influences that?
what is the TTL in the real case scenario? can this happen because the webserver is overloaded? I am trying to understand if this can translate to a real world issue. and if there are consequences?

obviously you hit an issue by setting the TTL so low that the CI computer cannot handle it. I am just wondering if silencing the exception is the way to go. 1/10th of a second is pretty small to context switch between 250 mutexes on the same lock.

Reply:
So I'm adding some new alerts for this to monitor the situation on production deployments